From 43b0c1af78d88a2efac05b5b0234307dcdf55ac2 Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Fri, 11 Oct 2024 10:38:29 +0200
Subject: [PATCH 1289/1695] plat-stm32mp1: scmi_server: permit MCU reset upon
 remoteproc security

Forbid SCMI accesses to MCU reset controllers when remote processor
is to be managed through OP-TEE remoteproc services.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 8037a6b6ecb49e1d4112c8f9c0c3713ba413b43b plat-stm32mp1: scmi_server: permit MCU reset upon remoteproc security]
Change-Id: Ib208f040df59632109624f50cecd8c1b03d59c41
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/410658
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
---
 core/arch/arm/plat-stm32mp1/scmi_server.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/core/arch/arm/plat-stm32mp1/scmi_server.c b/core/arch/arm/plat-stm32mp1/scmi_server.c
index 2700eddbc..9ae82d810 100644
--- a/core/arch/arm/plat-stm32mp1/scmi_server.c
+++ b/core/arch/arm/plat-stm32mp1/scmi_server.c
@@ -15,6 +15,7 @@
 #include <drivers/stm32_cpu_opp.h>
 #include <drivers/stm32mp_dt_bindings.h>
 #include <drivers/stm32_etzpc.h>
+#include <drivers/stm32_remoteproc.h>
 #include <initcall.h>
 #include <keep.h>
 #include <kernel/pm.h>
@@ -589,8 +590,14 @@ int32_t plat_scmi_rd_autonomous(unsigned int channel_id, unsigned int scmi_id,
 	assert(rd->rstctrl);
 
 #ifdef CFG_STM32MP15
+	/* Reset cycle on MCU hold boot is not supported */
 	if (rd->reset_id == MCU_HOLD_BOOT_R)
 		return SCMI_NOT_SUPPORTED;
+
+	/* Remoteproc driver may handle all MCU reset controllers */
+	if (rd->reset_id == MCU_R &&
+	    stm32_rproc_is_secure(STM32MP1_M4_RPROC_ID))
+		return SCMI_DENIED;
 #endif
 
 	/* Supports only reset with context loss */
@@ -623,6 +630,13 @@ int32_t plat_scmi_rd_set_state(unsigned int channel_id, unsigned int scmi_id,
 	if (id != STM32MP1_ETZPC_MAX_ID && stm32_etzpc_check_ns_access(id))
 		return SCMI_DENIED;
 
+#ifdef CFG_STM32MP15
+	/* Remoteproc driver may handle all MCU reset controllers */
+	if ((rd->reset_id == MCU_HOLD_BOOT_R || rd->reset_id == MCU_R) &&
+	    stm32_rproc_is_secure(STM32MP1_M4_RPROC_ID))
+		return SCMI_DENIED;
+#endif
+
 	assert(rd->rstctrl);
 
 	if (assert_not_deassert) {
-- 
2.47.2

